### PR TITLE
Use image ID in ps if tag has been updated

### LINKED ISSUE #12595: Docker API does not report correct images for /containers/json endpoint
# Description of problem

Docker API (`v.1.18`) does not report correct images for `/containers/json` endpoint. Instead of the actual image ID for an image not longer tagged the previous tag is reported.

# `docker version`

```
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): linux/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64
```

# `docker info`

```
Containers: 6
Images: 139
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 151
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-49-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 8
Total Memory: 31.29 GiB
Name: Ubuntu-1404-trusty-64-minimal
ID: QE6C:ZNQX:BADL:4BTY:7AK5:274U:7B7N:EUS7:FGYI:TE5Q:JPXI:FBK6
WARNING: No swap limit support
```

# `uname -a`

```
Linux Ubuntu-1404-trusty-64-minimal 3.13.0-49-generic #83-Ubuntu SMP Fri Apr 10 20:11:33 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

# Environment details (AWS, VirtualBox, physical, etc.):

Bare-Metal, Ubuntu, latest lxc-docker

# How reproducible

100%

# Steps to Reproduce:
1. `docker pull` an image
2. `docker run` it (and give the container a name)
3. Update the image
4. `docker pull` the changed image
5. Ask the `/containers/json` endpoint about the container

# Actual Results

- The image ID for the tag has changed
- The old image is shown with "Repository" `<none>` 
- The API reports the old tag for the image the container is running with

# Expected Results

- The image ID for the tag has changed
- The old image is shown with "Repository" `<none>` 
- The API reports the **image ID** for the image the container is running with

# Additional info

```
# curl 127.0.0.1:21/containers/json?all=1
[{"Command":"./gobuilder","Created":1429599867,"Id":"dc520983d708749528b900118383ac20143a657a156c5efadc678892d24215e3","Image":"luzifer/gobuilder-frontend:latest","Labels":{},"Names":["/gobuild-frontend"],"Ports":[{"IP":"0.0.0.0","PrivatePort":3000,"PublicPort":1060,"Type":"tcp"}],"Status":"Up 5 hours"}]
# docker inspect gobuild-frontend | grep -i image
        "Image": "luzifer/gobuilder-frontend:latest",
    "Image": "b8e06e06802772cad27d271eac378eca042ebac6e3653bea21bd0751e1f30dfa",
# docker images
REPOSITORY                   TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
luzifer/gobuilder-frontend   latest              b96f5f028d3c        2 hours ago         592 MB
<none>                       <none>              b8e06e068027        5 days ago          591.7 MB
```

### PR DESCRIPTION
**Pull request by thieman**
Fixes #12595.

Signed-off-by: Travis Thieman travis.thieman@gmail.com

### CODE REVIEW

File: daemon/list.go [hunk 1]
```diff
@@ -133,7 +133,18 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
 			ID:    container.ID,
 			Names: names[container.ID],
 		}
-		newC.Image = container.Config.Image
+
+		img, err := daemon.Repositories().LookupImage(container.Config.Image)
+		if err != nil {
+			// If the image can no longer be found by its original reference,
+			// it makes sense to show the ID instead of a stale reference.
+			newC.Image = container.ImageID
```
Comment (thieman) [inline daemon/list.go]: There's a note in `LookupImage` about how they want it to return `nil, nil` if the image can't be found. The current behavior is that errors bubble up, though, necessitating this scary-looking exception swallowing.

Comment (thieman) [inline daemon/list.go]: Arguably, this could be future-proofed for a post-`nil` world by adding an `img == nil` check. Would that be preferable?

Comment (LK4D4) [inline daemon/list.go]: I'm pretty sure that it's returning `err != nil` if image not found.

File: integration-cli/docker_cli_ps_test.go [hunk 2]
```diff
@@ -666,3 +666,57 @@ func (s *DockerSuite) TestPsGroupPortRange(c *check.C) {
 	}
 
 }
+
+// Test for GitHub issue #12595
+func (s *DockerSuite) TestPsImageIDAfterUpdate(c *check.C) {
+
+	runCmd := exec.Command(dockerBinary, "tag", "busybox:latest", "busybox:TestPsImageIDAfterUpdate-original")
```
Comment (LK4D4) [inline integration-cli/docker_cli_ps_test.go]: Maybe makes sense to do "busybox:TestPsImageIDAfterUpdate-original" variable.

File: integration-cli/docker_cli_ps_test.go [hunk 3]
```diff
@@ -666,3 +666,57 @@ func (s *DockerSuite) TestPsGroupPortRange(c *check.C) {
...
+	out, _, err := runCommandWithOutput(runCmd)
+	if err != nil {
+		c.Fatal(out, err)
+	}
+
+	runCmd = exec.Command(dockerBinary, "inspect", "-f", "{{ .Id }}", "busybox:TestPsImageIDAfterUpdate-original")
```
Comment (LK4D4) [inline integration-cli/docker_cli_ps_test.go]: We have utility function `inspectField`.

File: daemon/list.go (hunk 1 above)
Comment (relistan) [inline daemon/list.go]: I'm not sure you need to call `LookupImage`. I set up three containers and printed out the values of `Container.ImageID` and `Container.Config.Image`. I think by comparing these you can determine the correct result:

This one was invoked with no tag and the image name still matches the correct image.
```
Container.image 8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55
Container.Config.image busybox
```

This one has had the tag moved:
```
Container.ImageID 2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d
Container.Config.Image 2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d
```

Here the image name still points to the right image and the container was invoked with a tag:
```
Container.ImageID ec64a324e9ccfbd431deed183c2a34cd3ded0704d763378d71aebbf08c483f73
Container.Config.Image darron/centurion-test:latest
```

Comment (thieman) [inline daemon/list.go]: @relistan I'm still new to this, so please don't interpret this as me being snarky. Here are the two edge cases I ran up against when trying to get this to work, could we make these work without doing `LookupImage`?

1. Intermediate build containers where `container.Config.Image == ""`. This was causing me problems in `TestBuildContextCleanupFailedBuild`.
1. Situations in which a container is created using repo:tag and then that tag is deleted. Failing test for me here was `TestRmContainerOrphaning`.

Happy to be educated. :)

Comment (thieman) [inline daemon/list.go]: @LK4D4 That is the current behavior, yes.

Comment (relistan) [inline daemon/list.go]: No worries, I'm learning this stuff on the fly. You're ahead of me I think. :) I haven't gotten through a complete test suite run so I trust your results.

Comment (relistan) [inline daemon/list.go]: Completing my code and will run the suite. I'll report back if I get any different results in a bit.

Comment (relistan) [inline daemon/list.go]: Just noticed that I missed a question in your response @thieman. The second example you ask about is what is stored internally. The problem is that the code was still applying ":latest" because the image was originally started as "nginx:latest" on the command line.  Here's the actual Docker 1.6.0 output from that same container:

```
d419fa7ad1a7        2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d:latest   "nginx -g 'daemon of   9 weeks ago         Up 8 seconds        443/tcp, 0.0.0.0:10234->80/tcp   dockercon-6c01869525db08
```

Comment (thieman) [inline daemon/list.go]: Hmm, ok. Now I'm not quite sure I've fully understood the problem. Eager to check out what you've got.

Comment (relistan) [inline daemon/list.go]: I don't have any of your tests merged yet, about to run a full test suite. But [this commit](https://github.com/relistan/docker/tree/fixing-tags) seems to work from the command line anyway. Not pretty.

File: integration-cli/docker_cli_ps_test.go [hunk 4]
```diff
@@ -666,3 +666,57 @@ func (s *DockerSuite) TestPsGroupPortRange(c *check.C) {
...
 
 }
+
+// Test for GitHub issue #12595
+func (s *DockerSuite) TestPsImageIDAfterUpdate(c *check.C) {
+
+	runCmd := exec.Command(dockerBinary, "tag", "busybox:latest", "busybox:TestPsImageIDAfterUpdate-original")
+	out, _, err := runCommandWithOutput(runCmd)
+	if err != nil {
+		c.Fatal(out, err)
```
Comment (LK4D4) [inline integration-cli/docker_cli_ps_test.go]: Also would you mind to change `Fatal`s to `Assert`s?

**Comment by LK4D4** [comment]
@thieman Also note, that you should post your `Fixes ...` in top comment of PR, then github will be able to automatically close issue.

**Comment by LK4D4** [comment]
@thieman @relistan I think that probably this is not right way to do it. I think we should update `config.Image` if we doing something with that `config.Image`. WDYT?

**Comment by relistan** [comment]
@lk4d4 Personally I never found the original behavior to be problematic. It was consistent and it has been like that for a long time. The new behavior is actually worse because it's not just going to be the current state, it's going to either be what I asked for when I started the container, or it will be what I computed now because that original state is lost.  I much prefer a consistent answer.

But, if we have to work with the expectation that we want to see the tag we used in the `docker run` command, then we're constrained to doing some funny things here.

**Comment by LK4D4** [comment]
@relistan Original behavior had bug. If you have tags busybox:a and busybox:b and you created your container with busybox:b in "old"-ps you will see busybox:a, so, it's not consistent. I agree that with fixing that brought even worse bug. Our task here is to fix both.
So, problem is that `Config.Image` is pretty useless because it can be tag.

File: integration-cli/docker_cli_ps_test.go [hunk 5]
```diff
@@ -666,3 +666,58 @@ func (s *DockerSuite) TestPsGroupPortRange(c *check.C) {
 	}
 
 }
+
+// Test for GitHub issue #12595
+func (s *DockerSuite) TestPsImageIDAfterUpdate(c *check.C) {
+
+	originalImageName := "busybox:TestPsImageIDAfterUpdate-original"
+	updatedImageName := "busybox:TestPsImageIDAfterUpdate-updated"
```
Comment (relistan) [inline integration-cli/docker_cli_ps_test.go]: @thieman reading through this test I'm having a hard time telling if it's actually testing the problem I see. There could be more than one here. These are the steps that currently result in an error for us:

1. Image exists as `busybox`
2. I start up a container
3. `docker ps` shows my image as `busybox:latest`
3. `busybox` is outdated and I do a `docker pull` which updates the image
4. `docker ps` shows my image ID as *2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d:latest*

**Comment by relistan** [comment]
@lk4d4 Yes, the old behavior showed you something other than what you typed, but it was always the first tag, so that's what I meant by consistent. Yes, I'd like to fix both.

Great, if there is a reasonable way to do this farther up the chain, I agree that seems better. :) It's above my skill set at the moment, though.

Should I go ahead and prepare a PR with my alternate fix? I'm headed for bed now, but can do it tomorrow. If nobody wants to go down that road, though, I'll save everyone the work.

**Comment by relistan** [comment]
Hmm, the problem with fixing it farther up, when we move the image name, is that upgraded Docker installs (e.g. from 1.5.0 -> 1.6.0) will still exhibit broken output. That would be all of our production network.

File: integration-cli/docker_cli_ps_test.go (hunk 5 above)
Comment (thieman) [inline integration-cli/docker_cli_ps_test.go]: I was attempting to replicate jperville's test case from the issue, as well as Luzifer's original expected/actual report: https://github.com/docker/docker/issues/12595

Their issue looks different to me than what you and didip go on to report later in the thread. I think I need to do some poking around with a client to reconcile the two in my head.

**Comment by relistan** [comment]
@LK4D4 @thieman, I believe there are perhaps two issues here and I described them on #12595. I think the one @thieman is working on (I called "1") is probably best solved by @lk4d4's suggestion of fixing the image ID when we commit the container. The issue I called "2" I believe is a display issue that should be fixed in this code.

**Comment by LK4D4** [comment]
@relistan So, here we're talking about display issue in ps and events. But I think it might be related to updating issue as well.

File: daemon/list.go [hunk 6]
```diff
@@ -133,7 +133,18 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
 			ID:    container.ID,
 			Names: names[container.ID],
 		}
-		newC.Image = container.Config.Image
+
+		img, err := daemon.Repositories().LookupImage(container.Config.Image)
```
Comment (LK4D4) [inline daemon/list.go]: This logic also should be used in `events`. makes sense to create function.

**Comment by LK4D4** [comment]
ping @docker/core-maintainers 
WDYT about design?

File: daemon/list.go [hunk 7]
```diff
@@ -133,7 +133,18 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
...
+		if err != nil {
+			// If the image can no longer be found by its original reference,
+			// it makes sense to show the ID instead of a stale reference.
+			newC.Image = container.ImageID
+		} else if container.ImageID == img.ID {
+			newC.Image = container.Config.Image
+		} else {
+			newC.Image = container.ImageID
+		}
```
Comment (hqhq) [inline daemon/list.go]: Need to consider if newC.Image has no tag, then should we add the default tag as it used to do?

File: integration-cli/docker_cli_ps_test.go [hunk 8]
```diff
@@ -666,3 +666,44 @@ func (s *DockerSuite) TestPsGroupPortRange(c *check.C) {
...
+	c.Assert(err, check.IsNil)
+
+	originalImageID, err := getIDByName(originalImageName)
+	c.Assert(err, check.IsNil)
+
+	runCmd = exec.Command(dockerBinary, "run", "-d", originalImageName, "top")
+	out, _, err = runCommandWithOutput(runCmd)
+	c.Assert(err, check.IsNil)
+	containerID := strings.TrimSpace(out)
+
```
Comment (dmcgowan) [inline integration-cli/docker_cli_ps_test.go]: Can you add a check here that ps will return with originalImageName, this makes the last check at the end more valuable.

I tried with this and it passed 
```diff
diff --git a/integration-cli/docker_cli_ps_test.go b/integration-cli/docker_cli_ps_test.go
index 80bd1db..8baa5a7 100644
--- a/integration-cli/docker_cli_ps_test.go
+++ b/integration-cli/docker_cli_ps_test.go
@@ -689,14 +689,27 @@ func (s *DockerSuite) TestPsImageIDAfterUpdate(c *check.C) {
        out, _, err = runCommandWithOutput(runCmd)
        c.Assert(err, check.IsNil)
 
+       linesOut, err := exec.Command(dockerBinary, "ps").CombinedOutput()
+       c.Assert(err, check.IsNil)
+
+       lines := strings.Split(strings.TrimSpace(string(linesOut)), "\n")
+       // skip header
+       lines = lines[1:]
+       c.Assert(len(lines), check.Equals, 1)
+
+       for _, line := range lines {
+               f := strings.Fields(line)
+               c.Assert(f[1], check.Equals, originalImageName)
+       }
+
        runCmd = exec.Command(dockerBinary, "tag", "-f", updatedImageName, originalImageName)
        out, _, err = runCommandWithOutput(runCmd)
        c.Assert(err, check.IsNil)
 
-       linesOut, err := exec.Command(dockerBinary, "ps").CombinedOutput()
+       linesOut, err = exec.Command(dockerBinary, "ps").CombinedOutput()
        c.Assert(err, check.IsNil)
 
-       lines := strings.Split(strings.TrimSpace(string(linesOut)), "\n")
+       lines = strings.Split(strings.TrimSpace(string(linesOut)), "\n")
        // skip header
        lines = lines[1:]
        c.Assert(len(lines), check.Equals, 1)
```

**Comment by dmcgowan** [comment]
:+1: on design

**Comment by LK4D4** [comment]
@thieman Would you mind to rebase?

**Comment by thieman** [comment]
@LK4D4 Went ahead and rebased off of current master. Did you also want me to go ahead and squash these?

**Comment by duglin** [comment]
@thieman yes please

**Comment by thieman** [comment]
Squashed, cheers

**Comment by thieman** [comment]
The Windows testing seems really sad right now (last 20 or so jobs have all failed).

**Comment by tiborvass** [comment]
@thieman rebase please, thanks :)

**Comment by thieman** [comment]
Rebased, cheers

File: integration-cli/docker_cli_ps_test.go [hunk 9]
```diff
@@ -554,3 +554,44 @@ func (s *DockerSuite) TestPsFormatHeaders(c *check.C) {
...
+	runCmd = exec.Command(dockerBinary, "commit", containerID, updatedImageName)
+	out, _, err = runCommandWithOutput(runCmd)
+	c.Assert(err, check.IsNil)
+
+	runCmd = exec.Command(dockerBinary, "tag", "-f", updatedImageName, originalImageName)
+	out, _, err = runCommandWithOutput(runCmd)
+	c.Assert(err, check.IsNil)
+
+	linesOut, err := exec.Command(dockerBinary, "ps").CombinedOutput()
```
Comment (cpuguy83) [inline integration-cli/docker_cli_ps_test.go]: Let's actually do a --no-trunc here, because there's currently a bug where the full image id is shown but it should only show up to 12 chars... or in your assert below do `originalImageID[12]`

**Comment by cpuguy83** [comment]
One nit since the test is relying on a bug with the image ID output on the CLI.
Other than that LGTM.

**Comment by icecrime** [comment]
LGTM, but I'd like a final look by @dmcgowan or @stevvooe?

**Comment by thieman** [comment]
Rebased and updated to add `--no-trunc` per Brian's comment

**Comment by stevvooe** [comment]
LGTM, pending the addressing of @dmcgowan comments.

**Comment by cpuguy83** [comment]
@thieman are you able to make the requested change?
Also, will need a rebase.

**Comment by thieman** [comment]
@cpuguy83 I can try to get this done tomorrow before I enter unavailability for 2 weeks (hooray, vacation). I'm not sure what the etiquette is in terms of finishing other people's PRs, but I'm completely fine with someone else putting the finishing touches on it and striking my name from the commit log if it fixes the issue. I'm sorry if I'm blocking anything.

**Comment by thaJeztah** [comment]
@thieman would be great if you could rebase before you're leaving.

If there are still finishing touches after that, a maintainer can "carry" your commit; in that case, we will keep your original commit (and sign-off), and add a new commit with the finishing touches, so that your commit will still end up in the merge (and your name ends up in the history / contributors list) :+1: 

Thanks for letting us know you won't be available after tomorrow, and enjoy your vacation!

**Comment by thieman** [comment]
Rebased, added dmcgowan's test for the original image name, and squashed.

**Comment by dmcgowan** [comment]
LGTM

**Comment by cpuguy83** [comment]
LGTM
